chimera/precondition_evaluator.py

```diff
@@ -55,6 +56,33 @@ def __init__(self, sf_context, chimera_config_filepath, pge_config_filepath, set
                 file_name = '~/verdi/etc/settings.yaml'
             raise RuntimeError("Could not read settings file '{}': {}".format(file_name, e))
 
+    def repl_val_in_dict(self, d, val, job_params, root=None):
+        """
+        Recursive function to replace occurences of val in a dict with values from the job_params.
+        """
+
+        if root is None: root = []
+        matched_keys = []
+        for k, v in d.items():
+            rt = copy.copy(root)
+            rt.append(k)
+            if isinstance(v, dict):
+                matched_keys.extend(self.repl_val_in_dict(v, val, job_params, rt))
+            if v == val:
+                jp_key = '.'.join(rt)
+                # use job_params with explicit dot notation
+                if jp_key in job_params:
+                    d[k] = job_params[jp_key]
+                    matched_keys.append(jp_key)
+                # maintain backwards-compatibility of using job_param values without dot notation
+                elif k in job_params:
+                    d[k] = job_params[k]
+                    matched_keys.append(k)
+                else:
+                    logger.error("job_params: {}".format(json.dumps(job_params, indent=2, sort_keys=True)))
```

Comment:
Seems like over here, we need to check if the metadata field is part of the optional fields list. If it is not, then throw the error like it is implemented here. Otherwise, replace with an empty string and continue. For debugging purposes, maybe we can throw an info messages indicating that the metadata field couldn't be replaced so we're setting it to an empty string since it was detected to be optional?

Reply:
@mcayanan: thanks for clarifying the optional_fields feature. I've updated the branch with your suggestions.